### PR TITLE
REVSDL-2374 VR plugin frame

### PR DESCRIPTION
**Pull request by VladSemenyuk**

### CODE REVIEW

File: src/components/vr_module/include/vr_module/event_engine/event.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+#define SRC_COMPONENTS_VR_MODULE_INCLUDE_VR_MODULE_EVENT_ENGINE_EVENT_H_
+
+#include <string>
+
+namespace event_engine {
+
+template<typename EventMessage, typename EventID>
+class Event {
+ public:
+  /*
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event.h]: `/*` -> `/**` Doxygen comment should start from double stars.

File: src/components/vr_module/include/vr_module/event_engine/event.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+template<typename EventMessage, typename EventID>
+const EventMessage& Event<EventMessage, EventID>::event_message() const {
+  return event_message_;
+}
+
+template<typename EventMessage, typename EventID>
+Event<EventMessage, EventID>::Event(const EventMessage& message, const EventID& id)
+: event_message_(message),
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event.h]: Incorrect indent. 4 spaces.

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                    EventObserver<EventMessage, EventID>* const observer);
+
+  /**
+   * @brief Unsubscribes the observer from specific event
+   *
+   * @param event_id    The event ID to unsubscribe from
+   * @param observer    The observer to be unsubscribed
+   */
+  void remove_observer(const EventID& event_id,
+                    const EventObserver<EventMessage, EventID>* const observer);
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: Wrong indent.

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+   */
+  virtual ~EventDispatcher();
+
+
+  FRIEND_BASE_SINGLETON_CLASS(EventDispatcher);
+
+  // Data types section
+  typedef std::list<EventObserver<EventMessage, EventID>*> ObserverList;
+  typedef std::map<int32_t, ObserverList>      ObserversMap;
+  typedef std::map<EventID, ObserversMap>      EventObserverMap;
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: Typedef should be the first according with [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order)

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+  // Members section
+  sync_primitives::Lock                               state_lock_;
+  EventObserverMap                                    observers_;
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: You know I don't like few spaces in anything :)

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      (*observers)->on_event(event);
+  }
+}
+
+
+template<typename EventMessage, typename EventID>
+void EventDispatcher<EventMessage, EventID>::add_observer(
+                   const EventID& event_id,
+                   int32_t correlation_id,
+                   EventObserver<EventMessage, EventID>* const observer) {
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: I'm not sure that indent of these lines is correct.

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  observers_[event_id][correlation_id].push_back(observer);
+}
+
+
+template<typename EventMessage, typename EventID>
+void EventDispatcher<EventMessage, EventID>::remove_observer(
+                   const EventID& event_id,
+                   const EventObserver<EventMessage, EventID>* const observer) {
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: I'm not sure that indent of these lines is correct.

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        ++observer_it;
+      }
+    }
+  }
+}
+
+
+template<typename EventMessage, typename EventID>
+void EventDispatcher<EventMessage, EventID>::remove_observer(
+                  const EventObserver<EventMessage, EventID>* const observer) {
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: I'm not sure that indent of this line is correct.

File: src/components/vr_module/include/vr_module/event_engine/event_observer.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+template<typename EventMessage, typename EventID>
+class EventObserver {
+ public:
+  // Typedef for possible Observer ID's from mobile_apis functionID enum
+  typedef int64_t ObserverID;
+
+  /*
+   * @brief Constructor
+   *
+   */
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_observer.h]: I see that is constructor. I think this comment is useless. I would rather delete it.

File: src/components/vr_module/include/vr_module/event_engine/event_observer.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  EventObserver();
+
+  /*
+   * @brief Destructor
+   */
+  virtual ~EventObserver();
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_observer.h]: I see that is destructor. I think this comment is useless. I would rather delete it.

File: src/components/vr_module/include/vr_module/event_engine/event_observer.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+class EventObserver {
+ public:
+  // Typedef for possible Observer ID's from mobile_apis functionID enum
+  typedef int64_t ObserverID;
+
+  /*
+   * @brief Constructor
+   *
+   */
+  EventObserver();
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/event_engine/event_observer.h]: I see that is constructor. I think this comment is useless. I would rather delete it.

File: src/components/vr_module/include/vr_module/vr_module_event.h [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    functional_modules::MobileFunctionID> {
+ public:
+  /**
+   * @brief Constructor with parameters
+   *
+   * @param id Event ID. (mobile function name)
+   * @param message Params in mobile response
+   */
+  VRModuleEvent(const application_manager::MessagePtr& message, // TODO(VS): MessagePtr will be replaced with gpb generated class
+             const functional_modules::MobileFunctionID& id);
```
Comment (bonewell) [inline src/components/vr_module/include/vr_module/vr_module_event.h]: Wrong indent.

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+  }
+
+  // Call observers
+  typename ObserverList::iterator observers =  list.begin();
+  for (; list.end() != observers; ++observers) {
+      (*observers)->on_event(event);
+  }
+}
+
+
```
Comment (DKlimenko) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: please remove empty line

File: src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+template<typename EventMessage, typename EventID>
+void EventDispatcher<EventMessage, EventID>::add_observer(
+                   const EventID& event_id,
+                   int32_t correlation_id,
+                   EventObserver<EventMessage, EventID>* const observer) {
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  observers_[event_id][correlation_id].push_back(observer);
+}
+
+
```
Comment (DKlimenko) [inline src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h]: please remove empty line

**Comment by DKlimenko** [comment]
reviewed

**Pull request merged**